### ISSUE TITLE
The handling of static shapes in `IfElse` may be stricter than absolutely necessary

### ISSUE THREAD

**Issue by dfm** [aesara-devs/aesara#985]
The Aesara `ifelse` is stricter about shapes than Theano was. For example, the following works in Theano:

```python
import theano
import theano.tensor as tt
from theano.ifelse import ifelse

x = tt.scalar()
res = ifelse(tt.gt(x, 0.5), tt.zeros(3), tt.zeros(4))
theano.function([x], res)(2.0)
```

This returns `array([0., 0., 0.])` as expected.

But the equivalent Aesara code crashes with either

- `TypeError: IfElse requires compatible types for true and false return values`, or
- `Dimension 0 in Rebroadcast's input was supposed to be 1 (got 3 instead)` if one of the branches is broadcastable.

```python
import aesara
import aesara.tensor as at
from aesara.ifelse import ifelse

x = at.scalar()
res = ifelse(at.gt(x, 0.5), at.zeros(3), at.zeros(4))
aesara.function([x], res)(2.0)
```

To get the `Rebroadcast` error, change `at.zeros(4)` to `tt.zeros(1)`, for example.

This is not a deal breaker for me since I'm already supporting a JAX interface that has the same limitation, but @brandonwillard suggests that it may be possible to handle shapes more flexibly so I wanted to open this here!

## Versions and main components

* Aesara version: 2.7.1
* Aesara config (`python -c "import aesara; print(aesara.config)"`): On colab
* Python version: 3.7
* Operating system: Colab
* How did you install Aesara: (conda/pip) pip

**Comment by brandonwillard** [comment]
Now that a lot of our uses of `Rebroadcast` have been removed, this issue may have changed.  
It currently gives the same `TypeError`s during graph optimization, but does not fail to compile the graph or execute the compiled function:
```python
import aesara
import aesara.tensor as at
from aesara.ifelse import ifelse


x = at.scalar("x")
res = ifelse(at.gt(x, 0.5), at.zeros(3), at.zeros(4))

res_fn = aesara.function([x], res)
# ERROR (aesara.graph.opt): Optimization failure due to: cond_make_inplace
# ERROR:aesara.graph.opt:Optimization failure due to: cond_make_inplace
# ERROR (aesara.graph.opt): node: if{}(Elemwise{gt,no_inplace}.0, TensorConstant{(3,) of 0.0}, TensorConstant{(4,) of 0.0})
# ERROR:aesara.graph.opt:node: if{}(Elemwise{gt,no_inplace}.0, TensorConstant{(3,) of 0.0}, TensorConstant{(4,) of 0.0})
# ERROR (aesara.graph.opt): TRACEBACK:
# ERROR:aesara.graph.opt:TRACEBACK:
# ERROR (aesara.graph.opt): Traceback (most recent call last):
# ...
# TypeError: IfElse requires compatible types for true and false return values: true_branch=TensorType(float64, (3,)), false_branch=TensorType(float64, (4,))

res_fn(3.0)
# array([0., 0., 0.])
res_fn(-3.0)
# array([0., 0., 0., 0.])
```

**Comment by brandonwillard** [comment]
In general, I think that the `IfElse` `Op` contract and/or general graph assumptions say that the shapes of both branches should be exactly equal, so an error here may be necessary.

**Comment by dfm** [comment]
I think it's very sensible to require that both branches have the same type / shape / etc.!

**Comment by brandonwillard** [comment]
> In general, I think that the `IfElse` `Op` contract and/or general graph assumptions say that the shapes of both branches should be exactly equal, so an error here may be necessary.

I may have overstated the actual assumptions.

First off, different shapes are clearly supported in terms of `Op` evaluation (e.g. my example above), but the primary requirement is that the output `Type`s of each branch must at least conform in their dtypes and number of dimensions.

The assumptions I was referring to involve some old/vague comments in the code and the more recent incorporation of static shape information.  The latter simply hasn't been explicitly tested and/or given much (formal) consideration yet, so that's not saying much.

Of note is that `IfElse.infer_shape` produces branching shape results:
```python
import aesara
import aesara.tensor as at
from aesara.ifelse import ifelse


x = at.scalar("x")
res = ifelse(at.gt(x, 0.5), at.zeros(3), at.zeros(4))

res_fn = aesara.function([x], res.shape)

aesara.dprint(res_fn)
# MakeVector{dtype='int64'} [id A] 2
#  |if{shape,inplace} [id B] 1
#    |Elemwise{gt,no_inplace} [id C] 0
#    | |x [id D]
#    | |TensorConstant{0.5} [id E]
#    |TensorConstant{3} [id F]
#    |TensorConstant{4} [id G]

res_fn(3.0)
# array([3])
res_fn(-3.0)
# array([4])
```
In other words, `IfElse.infer_shape` actually *does* intentionally support different shape _values_ for each branch, but clearly not different shape _lengths_/dimensions.  I don't know about the rest of the related `IfElse` functionality and handling, though, but it could very well be that the only thing not properly supported yet is static shape information.
